keps/1337-preempt-within-cohort-while-borrowing/README.md

```diff
@@ -0,0 +1,484 @@
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### Flopping between preempting pair of workloads
+
+In a setup as described in (Story 1)[#story-1] consider there are two workloads,
+one in ClusterQueue A, and one in ClusterQueue B. There is not enough resources
+in the cohort to run both workloads at the same time.
+
+If the preemption is allowed for workloads of any priorities
+(as when `.preemption.reclaimWithinCohort` equals `Any`)
+then the following flopping happens: workload A preempts workload B, then p
+preempted workload B returns to the queue and preempts workload A to make progress.
+This repeats, and none of the workloads can make any progress.
+
+In order to mitigate this risk we only enable the preemption to reclaim
+quota within cohort, for workloads requiring borrowing when such preempted
+workloads are of lower priority. This behavior is enabled by setting
+`.preemption.borrowWithinCohort.policy: LowerPriority`.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### API
+
+```golang
+// ClusterQueuePreemption contains policies to preempt Workloads from this
+// ClusterQueue or the ClusterQueue's cohort.
+type ClusterQueuePreemption struct {
+  ...
+	// borrowWithinCohort provides configuration to allow preemption within
+	// cohort while borrowing. This configuration can only be enabled when
+	// reclaimWithinCohort is not `Never`.
+	BorrowWithinCohort *BorrowWithinCohort `json:"borrowWithinCohort,omitempty"`
+  ...
+}
+
+type BorrowWithinCohortPolicy string
+
+const (
+	BorrowWithinCohortPolicyNever         BorrowWithinCohortPolicy = "Never"
+	BorrowWithinCohortPolicyLowerPriority BorrowWithinCohortPolicy = "LowerPriority"
+)
+
+// BorrowWithinCohort contains configuration which allows to preempt workloads
+// within cohort while borrowing.
+type BorrowWithinCohort struct {
+	// policy determines the policy for preemption to reclaim quota within cohort while borrowing.
+	// Possible values are:
+	// - `Never` (default): do not reclaim quota within ClusterQueue while borrowing.
+	// - `LowerPriority`: allow preempting workloads to reclaim quota within the cohort,
+	//    but only by preempting lower-priority workloads.
+	//
+	// +kubebuilder:default=Never
+	// +kubebuilder:validation:Enum=Never;LowerPriority
+	Policy BorrowWithinCohortPolicy `json:"policy,omitempty"`
+
+	// maxPriorityThreshold allows to restrict the set of workloads for preemption
+	// to reclaim quota within the cohort, to only workloads with priority below
+	// or equal the specified level.
+	//
+	// +optional
+	MaxPriorityThreshold *int32 `json:"maxPriorityThreshold,omitempty"`
```

Comment:
```suggestion
	BelowPriority *int32 `json:"belowPriority,omitempty"`
```
WDYT @alculquicondor @tenzen-y @yaroslava-serdiuk ?
Shorter, also, it makes it "sharp" inequality, so I can in simply code check: `priority(candidate) < min(belowPriority, priority(preemptor))`, because `LowerPriority` is also sharp.

Reply:
Yeah, I think `MaxPriorityThreshold` makes sense, and we should not make API decisions just based on the implementation.

Reply:
But just to be clear this feature is about combination of `preemption` and `borrowing`. It enables preempting workloads even for borrowing workloads.

Reply:
> But just to be clear this feature is about combination of preemption and borrowing. It enables preempting workloads even for a borrowing workload.

That's right. I wanted to say that mixed semantics would be possible to cause confusion. It might be caused confusion to use some phrases having the aspect of preemption in `borrowWithinCohort` since we declare `borrow`.

Actually, I was confused phrase, `BorrowWithinCohort` in `ClusterQueuePreemption`. However, I couldn't find a better API name instead of `BorrowWithinCohort`.

Reply:
I see, I initially wanted to do `ReclaimWithinCohortConfig` like here: https://github.com/kubernetes-sigs/kueue/blob/53c6c54d77ca25b92720720e5417b42bc942abc8/keps/1337-preempt-within-cohort-while-borrowing/README.md#reclaimwithincohortconfig. But it is not ideal anyway, and lenghty.

Reply:
Note that this new `borrowWithinCohort` field is inside the `ClusterQueuePreemption` field, so it is directly related to preemption.

`maxPriorityThreshold` has `<=` semantics. I don't think the name has any problems and the implementation is not significantly harder than the alternative.